js/sri4node-attachments.js

```diff
@@ -1,1078 +1,1677 @@
+   * @returns {Promise<S3.HeadObjectCommandOutput>}
+   */
+  async function headFromS3(s3filename) {
+    debug(`get HEAD for ${s3filename}`);
 
-    async function getFileMeta(s3filename) {
-      let data = null;
-      try {
-        let result = await headFromS3(s3filename);
-        data = result;
-      } catch (ex) {
+    const params = { Bucket: fullPluginConfig.s3bucket, Key: s3filename };
 
-      }
-      return data;
+    // debug(params);
+    const awss3 = getAWSS3Client();
+    return await awss3.send(new S3.HeadObjectCommand(params));
+  }
+
+  /**
+   * Any error currently is silently discarded,
+   * so if this method returns null, it mans there was a problem
+   *
+   * @param {string} s3filename
+   * @returns {Promise<S3.HeadObjectCommandOutput>}
+   */
+  async function getFileMeta(s3filename) {
+    let data = null;
+    try {
+      const result = await headFromS3(s3filename);
+      data = result;
+    } catch (err) {
+      // silently discarded ?
+      debug(
+        `[getFileMeta] problem getting meta from S3 for ${s3filename}: ${err}`
+      );
     }
+    return data;
+  }
 
-    function downloadFromS3(outstream, filename) {
-      var deferred = Q.defer();
+  /**
+   *
+   * @param {TReadableStream} outstream
+   * @param {string} filename
+   * @returns {Promise<void>}
+   * @rejects {number | Error} if S3.send gives reply with http status code, we return the statuscode; else the error instance
+   */
+  async function downloadFromS3(outstream, filename) {
+    const { s3bucket } = fullPluginConfig;
+
+    const params = {
+      Bucket: s3bucket,
+      Key: filename,
+    };
 
-      var s3bucket = configuration.s3bucket;
-      var msg;
+    try {
+      const head = await headFromS3(filename);
+      if (head) {
+        const awss3 = getAWSS3Client();
 
-      let awss3 = createAWSS3Client();
-      var params = {
-        Bucket: s3bucket,
-        Key: filename
-      };
+        const response = await awss3.send(new S3.GetObjectCommand(params));
+        // const stream = awss3.getObject(params).createReadStream();
 
-      headFromS3(filename).then(function (exists) {
-        if (exists) {
+        // to be able to use transformToWebStream(), nodejs >= 18 is needed!
+        //  --> currently use response.Body with incorrect typescript types
+        // (see https://stackoverflow.com/a/67373050)
+        // const stream = response.Body.transformToWebStream();
+        const stream = response.Body;
 
-          const stream = awss3
-            .getObject(params)
-            .createReadStream()
+        return new Promise((resolve, reject) => {
+          // Also need to listen for close on outstream, to stop in case the request is aborted
+          // at client-side before the end of the input stream (S3 file).
+          outstream.on("close", () => {
+            debug("[downloadFromS3] stream closed prematurely");
+            reject(new Error("499 partial content"));
+          });
```

Comment:
Swap this block with the one at the bottom (`stream.pipe()`)?

Reply:
why?

Reply:
Because it respects the natural flow of handling the stream: first the normal scenario, then the possible exceptions. In that sense we should have: `pipe - end - error / close`.

Reply:
I prefer to connect the stream with pipe() as last to avoid a potential race condition where events could occur before a handler for the event is attached.